aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -1337,6 +1337,104 @@ void qadaptive_avg_pool2d_nhwc_kernel(
+    float output_scale = qy.q_scale();
+    int input_zero_point = qx.q_zero_point();
+    int output_zero_point = qy.q_zero_point();
+
+    for (int64_t od = 0; od < osizeD; od++) {
+      int istartD = (int)std::floor((float)(od * isizeD) / osizeD);
+      int iendD = (int)std::ceil((float)((od + 1) * isizeD) / osizeD);
+      int kD = iendD - istartD;
+      for (int64_t oh = 0; oh < osizeH; oh++) {
+        int istartH = (int)std::floor((float)(oh * isizeH) / osizeH);
+        int iendH = (int)std::ceil((float)((oh + 1) * isizeH) / osizeH);
+        int kH = iendH - istartH;
+        for (int64_t ow = 0; ow < osizeW; ow++) {
+          auto* o_p = reinterpret_cast<typename scalar_t::underlying*>(
+              odata + b * osizeD * osizeH * osizeW * sizeC +
+              (od * osizeH * osizeW + oh * osizeW + ow) * sizeC);
+          int istartW = (int)std::floor((float)(ow * isizeW) / osizeW);
+          int iendW = (int)std::ceil((float)((ow + 1) * isizeW) / osizeW);
+          int kW = iendW - istartW;
+          int size = kD * kH * kW;
+          float multiplier = input_scale / output_scale / size;
+          int input_zero_point_m_size = -input_zero_point * size;
+          int64_t c = 0;
+          // For int8 or uint8quantization, we implicitly use int32 as
+          // accumulation Or else, it will go to the slow path
+          // TODO: support 16bit, 32bit, and etc.
+          auto* internal_i_p = i_p + istartH * istrideH + istartW * istrideW;
+
+          // TODO: more vectorization with loop interleaving
+          do_avg_pool_on_AVX2<scalar_t>(
+              internal_i_p,
+              o_p,
+              c,
+              sizeC,
+              1,
+              input_zero_point_m_size,
+              output_zero_point,
+              multiplier,
+              0,
+              kD,
+              0,
+              kH,
+              0,
+              kW,
+              istrideC,
+              istrideD,
+              istrideH,
+              istrideW);
+          // 1) The following loop handles the remaining channels
+          // 2) It also handles the Non-AVX2 path
+          for (; c < sizeC; ++c) {
+            int32_t acc_int32 = input_zero_point_m_size;
+            int64_t tcntr = 0;
+            for (int64_t id = 0; id < kD; id++) {
+              for (int64_t ih = 0; ih < kH; ih++) {
+                for (int64_t iw = 0; iw < kW; iw++) {
+                  tcntr = id * istrideD + ih * istrideH + iw * istrideW;
+                  auto val = *(internal_i_p + tcntr + c * istrideC);
+                  acc_int32 += val;
+                }
+              }
+            }
+            // clamp
+            o_p[c] = at::native::quantize_val<scalar_t>(1.0f / multiplier,
+                                                        output_zero_point,
+                                                        acc_int32).val_;
+          } // c
+        } // oh
+      } // ow
+    } // od
+  });
+}
+
 void qavg_pool2d_nhwc_kernel(
```

Comment:
I didn't read this code line by line, but I'm guessing it's not easily reusable/extensable for the 3d case?

Reply:
I tried reusing it, but it gets really convoluted. I will make a low-pri task to reuse the code, which will require refactoring :(

Reply:
#40316

Reply:
I was looking into this a little more -- I could reuse it by only writing the 3d version of it, and when using the 2d version, just set the extra dimension to 1. The problem is that this introduces extra computation steps (like adding zeros or multiplying by 1). Do we need that?

Reply:
I think we can worry about code reuse later, we want to make it work first

Reply:
makes sense, thanks for the context!

optional suggestion, if the new func is a copy and modify of the existing `qavg_pool2d_nhwc_kernel` function, can we keep the relevant variable names consistent?  would make it easier to realize that it's a copy and to refactor later